core/src/main/scala/kafka/log/remote/TopicPartitionRemoteIndex.scala

```diff
@@ -0,0 +1,236 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.log.remote
+
+import java.io.{File, IOException}
+import java.nio.file.{Files, Path}
+import java.util.Comparator
+import java.util.concurrent.locks.ReentrantLock
+import java.util.concurrent.{ConcurrentNavigableMap, ConcurrentSkipListMap}
+import java.util.function.{Consumer, Predicate}
+
+import kafka.log.remote.TopicPartitionRemoteIndex.{REMOTE_OFFSET_INDEX_SUFFIX, REMOTE_TIME_INDEX_SUFFIX}
+import kafka.log.{CleanableIndex, Log, OffsetIndex, TimeIndex}
+import kafka.utils.{CoreUtils, Logging}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.utils.Utils
+
+import scala.collection.JavaConverters._
+
+private case class RemoteSegmentIndex(offsetIndex: OffsetIndex, timeIndex: TimeIndex, remoteLogIndex: RemoteLogIndex) {
+  def asList(): Seq[CleanableIndex] = {
+    List(offsetIndex, timeIndex, remoteLogIndex)
+  }
+}
+
+/**
+ * This class contains all the indexes related to remote log segments for a given topic partition.
+ *  - remote log index
+ *    - maintains a mapping between offset/timestamp of logs and the location of the remote segment where the data exists.
+ *  - offset index
+ *    - maintains a mapping between offset and position in the remote log index which contains remote log information for this offset.
+ *  - time index
+ *   - maintains a mapping between timestamp and position in the remote log index which contains remote log information for this timestamp.
+ *
+ *   These index files are also locally stored in respective log-dir of the topic partition.
```

Comment:
Can these index files be reconstructed in case they are corrupted for whatever reason, e.g., unclean shutdown?

Reply:
Right, these can be reconstructed using the below APIs. This is a pending `todo` as mentioned [here](https://github.com/apache/kafka/pull/7561/files/7c9abf3a3bca2024092bd823f3225dafc8e84bb9#diff-eeb791e578ced303c3141c5369366f80R103)

```
listRemoteSegments(topicPartition: TopicPartition, minBaseOffset: Long)
getRemoteLogIndexEntries(remoteLogSegment: RemoteLogSegmentInfo): util.List[RemoteLogIndexEntry]
```

Reply:
Do you plan to hook some sort of recovery of remote indexes in `remote.log.LogSegment#recover`? However, since only inactive segments are on the remote storage, I wonder to which extent they should be involved in the recovery process.